### ISSUE TITLE
Multigpu not working

### ISSUE THREAD

**Issue by DylanNoone** [bpowell122/tomodrgn#16]
**Describe the bug**
Trying to use multigpu to speed up learning heterogeneity.

**To Reproduce**
#!/bin/sh

#SBATCH --partition="bubeck"
#SBATCH --job-name=TomoDRGN
#SBATCH --output=TomoDRGN-%j.out
#SBATCH --error=TomoDRGN-%j.err
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=90
#SBATCH --time=336:00:00
#SBATCH --gres=gpu:a10:4

# Load CUDA
module purge
module load cuda/12.4

# Initialize conda
source /bbsrc/home/dnoone/soft/conda/etc/profile.d/conda.sh

# Activate the environment
conda activate tomodrgn

#Force GPU visibility - I had problems with this
export CUDA_VISIBLE_DEVICES=0,1,2,3
export PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True

# Command

# note: --lazy is used because the separate .mrcs file per particle, as used by WarpTools,
#       is well suited to lazy loading.
# note: --num-workers, --prefetch-factor, and --persistent-workers are best used only if
#       --lazy is enabled to avoid excessive memory utilization.

tomodrgn \
    train_vae \
    run2_optimisation_set.star \
    --outdir tomodrgn_output/vae_warptools_70S_dosetiltweightmask \
    --zdim 8 \
    --uninvert-data \
    --num-epochs 40 \
    --l-dose-mask \
    --recon-dose-weight \
    --recon-tilt-weight \
    --batch-size 16 \
    --lazy \
    --num-workers 4 \
    --prefetch-factor 2 \
    --multigpu \
    --persistent-workers
    
# ----------------------------
# Deactivate conda environment
# ----------------------------
conda deactivate

**Expected behavior**
I expected utilisation across my 4 gpus on my node. But i get this:

+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 570.211.01             Driver Version: 570.211.01     CUDA Version: 12.8     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A10                     On  |   00000000:4B:00.0 Off |                    0 |
|  0%   74C    P0            134W /  150W |    7406MiB /  23028MiB |      0%   E. Process |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA A10                     On  |   00000000:65:00.0 Off |                    0 |
|  0%   44C    P8             18W /  150W |      13MiB /  23028MiB |      0%   E. Process |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   2  NVIDIA A10                     On  |   00000000:B1:00.0 Off |                    0 |
|  0%   39C    P8             28W /  150W |      13MiB /  23028MiB |      0%   E. Process |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   3  NVIDIA A10                     On  |   00000000:CA:00.0 Off |                    0 |
|  0%   42C    P8             37W /  150W |      13MiB /  23028MiB |      0%   E. Process |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A            3859      G   /usr/libexec/Xorg                         4MiB |
|    0   N/A  N/A          524297      C   ...onda/envs/tomodrgn/bin/python       7390MiB |
|    1   N/A  N/A            3859      G   /usr/libexec/Xorg                         4MiB |
|    2   N/A  N/A            3859      G   /usr/libexec/Xorg                         4MiB |
|    3   N/A  N/A            3859      G   /usr/libexec/Xorg                         4MiB |
+-----------------------------------------------------------------------------------------+

Is this a known bug and is there a work around?

**Comment by jtcarrion** [comment]
Thank you so much for reporting this, I am sorry for the delayed response but after reading into the code this makes sense. It seems `train_vae` `--multigpu` wraps the model in `DataParallelPassthrough`, but the training code mostly bypasses `DataParallel.forward` as is. I should have this and some other small bugs fixed by the end of the weekend, thank you again for the very clean and detailed documentation!!
